### PR TITLE
add CategoryController, BookDtoWithoutCategoriesIds, CategoryDto, Cat…

### PR DESCRIPTION
**Pull request by Alexander1iubchenko**
add CategoryController, BookDtoWithoutCategoriesIds, CategoryDto, CategoryRequestDto, CategoryMapper, Category model, CategoryRepository, CategoryService, CategoryServiceImpl, 04-create-catrgories-table.yaml, 08-insert-categories.yaml,09-insert-books.yaml, 10-create-books-categories-table.yaml

### CODE REVIEW

File: src/main/java/model/Book.java [hunk 1]
```diff
@@ -41,6 +47,11 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (ponomvrenko) [inline src/main/java/model/Book.java]: This is bad practice, try to use @EntityGraph

File: src/main/java/repository/BookRepository.java [hunk 2]
```diff
@@ -1,9 +1,23 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.EntityGraph;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long> {
+    @Query("SELECT b FROM Book b JOIN FETCH b.categories c WHERE c.id = :categoryId")
```
Comment (ponomvrenko) [inline src/main/java/repository/BookRepository.java:15]: You can replace this query with EntityGraph

**Review by ponomvrenko — CHANGES_REQUESTED** [review]
Good job! Some small changes are need to be done.

File: src/main/java/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
+    public List<BookDtoWithoutCategoriesIds> getBooksByCategoryId(@PathVariable Long id) {
+        return bookService.getBooksByCategoryId(id);
+    }
+
+    @PostMapping
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Create a new category",
```
Comment (d1sam) [inline src/main/java/controller/CategoryController.java:59]:
```suggestion
    @ResponseStatus(HttpStatus.CREATED)
    @Operation(summary = "Create a new category",
```

File: src/main/java/model/Book.java [hunk 4]
```diff
@@ -41,6 +46,11 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany
+    @JoinTable(name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id"))
+    private Set<Category> categories = new HashSet<>();
```
Comment (d1sam) [inline src/main/java/model/Book.java]: This field should be excluded from ToString and Equals/HashCode

File: src/main/java/model/Category.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    public Category() {
+    }
```
Comment (d1sam) [inline src/main/java/model/Category.java]: replace with `@NoArgsConstructor` for better readability

File: src/main/java/model/Category.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "categories")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
```
Comment (d1sam) [inline src/main/java/model/Category.java:27]:
```suggestion
    @Column
    private String description;
```

File: src/main/resources/db/changelog/changes/04-create-categories-table.yaml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-roles-table
+      author: alex
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-categories-table.yaml]:
```suggestion
      author: alex_liubchenko
```
it's better to pass also last name

File: src/main/resources/db/changelog/changes/04-create-categories-table.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      changes:
+        - createTable:
+            tableName: categories
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    nullable: false
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-categories-table.yaml]:
```suggestion
```
you have primary key, that's why nullable constraint is unnecessary

File: src/main/resources/db/changelog/changes/08-insert-categories.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert_categories
+      author: alex
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/08-insert-categories.yaml]:
```suggestion
      author: alex_liubchenko
```
it's better to pass also last name

File: src/main/resources/db/changelog/changes/09-insert-books.yaml [hunk 10]
```diff
@@ -0,0 +1,68 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert_books
+      author: alex
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/09-insert-books.yaml]:
```suggestion
      author: alex_liubchenko
```
it's better to pass also last name

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 11]
```diff
@@ -6,8 +6,16 @@ databaseChangeLog:
...
-      file: db/changelog/changes/06-create-users-roles-table.yaml
+      file: db/changelog/changes/06-insert-users.yaml
+  - include:
+      file: db/changelog/changes/07-create-users-roles-table.yaml
+  - include:
+      file: db/changelog/changes/08-insert-categories.yaml
+  - include:
+      file: db/changelog/changes/09-insert-books.yaml
+  - include:
+      file: db/changelog/changes/10-create-books-categories-table.yaml
```
Comment (d1sam) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/10-create-books-categories-table.yaml
      
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but some changes should be done!

File: src/main/java/controller/CategoryController.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    @GetMapping("/{id}")
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
+    @Operation(summary = "Get a category by id", description = "Gets a category by id")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
+        return categoryService.getCategoryById(id);
+    }
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
+    public List<BookDtoWithoutCategoriesIds> getBooksByCategoryId(@PathVariable Long id) {
```
Comment (d1sam) [inline src/main/java/controller/CategoryController.java:51]: add pagination

**Review by d1sam — CHANGES_REQUESTED** [review]

**Review by d1sam — APPROVED** [review]
Well done!

**Pull request merged**